### ISSUE TITLE
hash mismatches and other problems

### ISSUE THREAD

**Issue by aaronovz1** [clearlinux/swupd-client#590]
Hello me again,

Still hammering away at the Yocto integration. I have mixer/docker integrated and building updates. I have now started testing the client side doing a simple update from version 1 to version 2. The update just adds one file to the system `/opt/seeingmachines/dss/version_test` in the `dss` bundle.

The problem I am seeing is that when doing `swupd verify --fix` it appears that every file on the system fails the hash check. I am not sure how best to track down the cause of this and was hoping to get some help on narrowing it down.

Can someone help me determine if the update data generated by mixer seems appropriate? I have uploaded the entire www directory [here](https://www.dropbox.com/s/s346q2nweu1frge/www.tar.xz?dl=0).

The log I'm getting from the update process is below. One thing to note is I am using this in an A/B partition setup using the [script](https://github.com/aaronovz1/meta-swupd/blob/master/recipes-core/swupd-client/files/swupd-update-partition.sh) @pohly wrote in meta-swupd. This script works well in our old environment that is still using swupd-server, so I'm leaning toward the issue being with the update generation in mixer.
```
...
[INFO] : Update target device: /dev/mmcblk1p4
[INFO] : Content URL: http://172.23.1.55/update
[INFO] : swupd-update-partition: Bind-mounting source tree.
[INFO] : swupd-update-partition: mount -obind,ro /media/rfs/ro /tmp/swupd-mount-source.1SuJKV
[INFO] : swupd-update-partition: Updating to 2 from http://172.23.1.55/update.
[INFO] : swupd-update-partition: Trying to update.
[INFO] : swupd-update-partition: mount /dev/mmcblk1p4 /tmp/swupd-mount.jH0CFU
[INFO] : swupd-update-partition: Copy from source /media/rfs/ro.
[INFO] : : Can't translate pathname './usr/share/ca-certificates/mozilla/NetLock_Arany_=Class_Gold=_Főtanúsítvány.crt' to UTF-8
[INFO] : : Can't translate pathname './usr/share/ca-certificates/mozilla/TÜBİTAK_UEKAE_Kök_Sertifika_Hizmet_Sağlayıcısı_-_Sürüm_3.crt' to UTF-8: Can't translate pathname './usr/share/ca-certificates/mozilla/TÜRKTRUST_Elektronik_Sertifika_Hizmet_Sağlayıcısı_H5.crt' to UTF-8: Can't translate pathname './usr/share/ca-certificates/mozilla/Certinomis_-_Autorité_Racine.crt' to UTF-8
[INFO] : swupd-update-partition: Trying to update.
[INFO] : swupd-update-partition: swupd update --no-scripts -c http://172.23.1.55/update -v file:///tmp/swupd-version.wYMlkV -S /tmp/swupd-mount.jH0CFU/swupd-state -p /tmp/swupd-mount.jH0CFU
[INFO] : Update started.
[INFO] : Preparing to update from 1 to 2
[INFO] : Downloading packs...
[INFO] : Extracting dss pack for version 2
[INFO] : Extracting os-core pack for version 2
[INFO] : Statistics for going from version 1 to version 2:
[INFO] : changed bundles   : 2
         new bundles       : 0
         deleted bundles   : 0

         changed files     : 2
         new files         : 1
         deleted files     : 0

[INFO] : Starting download of remaining update content. This may take a while...
[INFO] : Finishing download of update content...
[INFO] : Staging file content
[INFO] : Applying update
[INFO] : Update was applied.
[INFO] : WARNING: post-update helper scripts skipped due to --no-scripts argument
[INFO] : Update took 10.7 seconds
[INFO] : Update successful. System updated from version 1 to version 2
[INFO] : swupd-update-partition: Verifying and fixing content.
[INFO] : swupd-update-partition: swupd verify --fix --picky --picky-tree /media/rfs/ro --picky-whitelist /swupd-state --no-scripts -F 4 -c http://172.23.1.55/update -v file:///tmp/swupd-version.wYMlkV -m 2 -S /tmp/swupd-mount.jH0CFU/swupd-state -p /tmp/swupd-mount.jH0CFU
[INFO] : Verifying version 2
[INFO] : Verifying files
[INFO] : Starting download of remaining update content. This may take a while...
[INFO] : Warning: Linkname can't be converted from UTF-8 to current locale.
[INFO] : Warning: Linkname can't be converted from UTF-8 to current locale.
[INFO] : Warning: Linkname can't be converted from UTF-8 to current locale.
[INFO] : Warning: Linkname can't be converted from UTF-8 to current locale.
[INFO] : Finishing download of update content...
[INFO] : Adding any missing files
[INFO] : Fixing modified files
[INFO] : Hash mismatch for file: /tmp/swupd-mount.jH0CFU/bin
[INFO] : 	fixed
[INFO] : Hash mismatch for file: /tmp/swupd-mount.jH0CFU/bin/arping
[INFO] : 	fixed
[INFO] : Hash mismatch for file: /tmp/swupd-mount.jH0CFU/bin/ash
[INFO] : 	fixed
...
```

**Comment by aaronovz1** [comment]
The issue from #445 is present here.

I started digging through the manifests based on documentation [here](https://github.com/clearlinux/swupd-server/wiki/Manifest-changes). It does seem like the update generation is correct and it picked up that only a couple changes were done, which reflects the change I made. I'd still appreciate if anyone can confirm this.

I'm curious if somehow the hashes are being modified due to file attributes/xattr not being retained somewhere along the line. When I put the initial version 1 image onto the system, I am simply doing a `cp -a src dst` to copy the rootfs from the image onto the device. `-a` is meant to preserve everything so I wouldn't think the issue lies here?

Does anyone have any ideas on this? I feel like I'm so close to wrapping this up. Just needs some pointers on where to track it down.

**Comment by otaviobp** [comment]
One hint on how to debug this problem is to use swupd hashdump. If you are unsure if one operation will change the hash you can always run swupd hashdump before and after the operation. If you are unsure if the installed files are correct, you can check the swupd hashdump too. cp -a shoudn't affect the file hash, though.

**Comment by aaronovz1** [comment]
Thanks for the info. I had started looking into the hashes. Is my assumption correct that the hash from hashdump should match what is listed in the manifest? In that case the hash doesn't match and I think it likely because of how I'm doing the chroots for the bundles. I'm going to retest once I use the method suggested in https://github.com/clearlinux/mixer-tools/issues/476. I'll report back once I have some more info.

**Comment by otaviobp** [comment]
> Is my assumption correct that the hash from hashdump should match what is listed in the manifest?

yes

**Comment by aaronovz1** [comment]
Hi @otaviobp / @matthewrsj 
I'm struggling to track down this issue. I looked into hashdump, and some hashes match but many don't. For instance the hash for `/usr/bin/addbuiltin` is `1554e135da1c2e6b037739830726274ada7f95c445eb46e2678ae3a25b5efdd3` but in the manifests it is listed as `bb33a1995f0a186e7d0583df6246f34c4e653f4e8dd09b9df2c9b08376701b32`

However if I check `/home/dss/.profile` then the hash of `6e4069ad07fae9077ae7b823d2c5afcbd8b3208d2b6e390ba30d8f71c2df4b3f` matches in manifests and on the target.

I have attached my generated bundle-info's and manifests. Can someone take a look to see if there's something obvious I'm doing wrong?
[mixer-integration.zip](https://github.com/clearlinux/swupd-client/files/2512259/mixer-integration.zip)

Thanks

**Comment by aaronovz1** [comment]
I tried running it with and without and result was the same. I actually might have just figured out what is going on. After doing some further comparisons, I noticed that the files which had correct hashes were the ones with uid/gid the same as my build system (1000), and anything that was meant to be owned by root or anything other than 1000, was wrong.

I'm pretty sure the issue is that I'm not unpacking the full chroot using docker (sudo), but the host. This would mess up the uid/gid and thus meta -> hashes.
